Solving #11428: Simplifying the function validateVolumeTemplate

pat-glitch opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
Before this PR: 
func validateVolumeTemplate in the file vmsadmitter.go had multiple nesting levels and needed to be simplified.

After this PR:

Solves #11428, simplifies the validate dataVolumeTemplate function.

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->
Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

1. I made a check for an empty 'DataVolumeTemplates' slice at the beginning of the function to exit early if there are no templates to validate.
2. The validation checks for `Name` and  `dataVolumeRefernce` is performed sequentially without nested blocks.

Basically, using the early returns and a separate functions, have reduced the nesting levels in the main function.

The following alternatives were considered:

I considered separating the validating tests into their own functions, but I didn't know if it was the correct way forward. So took the above route. 

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
4. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

0xFelix commented on pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go:
```diff
@@ -356,41 +356,43 @@ func ValidateVirtualMachineSpec(field *k8sfield.Path, spec *v1.VirtualMachineSpe
 }
 
+func checkDataVolumeReference(field *k8sfield.Path, idx int, dataVolume v1.DataVolumeTemplateSpec, volumes []v1.Volume) bool {
+	for _, volume := range volumes {
```
Can you keep the TODO comment here please?

pat-glitch replied:
Yes definitely, can I try and solve it as well?

0xFelix replied:
What change do you propose? Not sure there is anything to change about this right now.

pat-glitch replied:
The current logic assusmes that the PVC name equals the DV name right, and according to the TODO statement it might not be the same case in the future, what if we introduce a 'map' like `pvctoDVmap` to store the mapping between PVC names and DV names.
Then when we are supposed to check for the presence of DVs in the 'Volumes' list, we consult this map to match the PVC names with DV names.

pat-glitch replied:
@0xFelix @alicefr How does that solution sound for the TODO comment?

0xFelix reviewed (commented):
Looking good, thanks! One comment and please add a DCO signoff and a `None` to the release note in the PR description.

pat-glitch commented:
Hey @0xFelix, have updated the PR with a comment, added a DCO signoff and added the NONE to the release note, please can you check.

0xFelix commented:
@pat-glitch Can you squash the commits to a single commit please?

pat-glitch commented:
Yes I'll do that

xpivarc commented:
/cc

alicefr commented:
@pat-glitch any reason for closing this?

pat-glitch commented:
Sorry, I didn't want to close this as completed, I wanted to close it and put in a new PR, with the corrections done in one merge, I tried the git squash command, but couldn't squash the merges into one, so was trying to open a new PR.

0xFelix commented:
There is no need to create new PRs. You can force push the branch in your fork until everything is fixed. Please re-open this PR.

pat-glitch commented:
I'm sorry @0xFelix, @alicefr, I tried this and it I could not squash it into one, in the end I deleted the fork with this PR (I deleted it yesterday), and started all over again, I have just raised a new PR (#11473) with everything done correctly. I'm really sorry everyone.

alicefr commented:
@pat-glitch itsn't too bad, it just creates a bit of noise for the reviewers. Just try to force push when you want to publish your update. If you delete remotely the branch from where you have created the PR, then this automatically close your PR too,

pat-glitch commented:
> @pat-glitch itsn't too bad, it just creates a bit of noise for the reviewers. Just try to force push when you want to publish your update. If you delete remotely the branch from where you have created the PR, then this automatically close your PR too.

I was using the rebase command with HEAD**-**(n) instead of HEAD**~**(n), sorted it out in the new PR, that's why it wasn't working properly.
